### ISSUE TITLE
Hide or move away from Span the dangerous APIs

### ISSUE THREAD

**Issue by VSadov** [dotnet/corefx#18747]
Every time I look through the Span API, the following leaves some bad taste:

```C#
public static Span<T> DangerousCreate(object obj, ref T objectData, int length)
```

The method is very unsafe, but the only protection is a scary name. 
Is it possible to make `length` a `void*` to force that caller uses "unsafe" context to call this method?

The method is here for reference:  https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Span.cs#L136

Another option is to move this and other dangerous methods to a separate location such as `System.Runtime.CompilerServices.Unsafe`.

**Comment by VSadov** [comment]
CC: @jkotas @KrzysztofCwalina @jaredpar

**Comment by jkotas** [comment]
> Is it possible to make length a void* to force that caller uses "unsafe" context to call this method?

I do think that it is a good design. Instead, we should introduce a custom attribute that will only allow use of the API in unsafe context.

**Comment by jkotas** [comment]
Also see:

https://github.com/dotnet/coreclr/issues/3143
https://github.com/dotnet/roslyn/issues/8663
https://github.com/dotnet/roslyn-analyzers/issues/972

**Comment by VSadov** [comment]
I have a feeling that https://github.com/dotnet/roslyn-analyzers/issues/972 will not be implemented before Span ships. Once shipped adding the attribute would be a breaking change. 

Also, relying on an optional analyzer as a way to protect type safety and prevent GC holes seem not reliable enough. 

I think, since it is right there and so easily discoverable, we need to do something.
The method is right there in the public API of `Span`, not even in some special `System.Unsafe` location, which users will learn to avoid or use with great care.

 - If I am doing a code review and someone uses Span.DangerousCreate, I will have no idea how really dangerous this API is. 
One thing not checking arguments and maybe failing later in some strange way, but resulting in a GC hole seems way worse outcome.

**Comment by jkotas** [comment]
This problem existed since forever. The framework design pattern used for it is to include Unsafe or Dangerous in the name and make it less discoverable. I agree that it would be nice to have better solution for it.

The unsafe APIs exist for performance reasons. Making them artificially slower by adding dummy arguments is contraproductive. Making them less discoverable sounds find to me. Note that DangerousCreate is not the only dangerous API on Span.

**Comment by VSadov** [comment]
I am not arguing for solving this problem in general, I know the problem exists.

I just think that we are really inviting problems here. It is like having a dangerous API defined right on a common primitive like `System.String`. We usually move that stuff to something like `S.R.CompilerServices` or `S.R.InteropServices` or now to `S.R.CompilerServices.Unsafe`.

I assumed that converting to `void*` and then back to `int` would be fairly cheap on an `AggressiveInlining` method. Most likely no-op. 
That would not work for `DangerousGetPinnableReference` though.

Perhaps we should move it to `S.R.CompilerServices.Unsafe`, together with `DangerousGetPinnableReference`?

**Comment by jkotas** [comment]
Moving it to `S.R.CompilerServices.Unsafe` would mean coupling System.Memory package with it. I do not think that it would work well.

Somewhere near sounds reasonable.

**Comment by VSadov** [comment]
`S.R.CompilerServices.UnsafeSpanOperations` ?

**Comment by jkotas** [comment]
This either need to be fixed for 2.1.0 or won't fixed. It won't be possible to fix this after 2.1.0.
